provider/reprovider.go

```diff
@@ -236,208 +302,223 @@ func initialReprovideDelay(duration time.Duration) Option {
+				maxCollectionDurationTimer.Stop()
+				goto ProcessBatch
+			case <-maxCollectionDurationTimer.C:
+				// If this timer has fired then the pause timer has started, so stop it.
+				pauseDetectTimer.Stop()
+				goto ProcessBatch
+			case <-s.ctx.Done():
+				return
 			}
-			pauseDetectTimer.Stop()
-			maxCollectionDurationTimer.Stop()
-		ProcessBatch:
+		}
 
-			if len(m) == 0 {
+		pauseDetectTimer.Stop()
+		maxCollectionDurationTimer.Stop()
+	ProcessBatch:
+		keys := make([]multihash.Multihash, 0, len(m))
+		for c := range m {
+			delete(m, c)
+			// hash security
+			if err := verifcid.ValidateCid(s.allowlist, c); err != nil {
+				log.Errorf("insecure hash in reprovider, %s (%s)", c, err)
 				continue
 			}
+			keys = append(keys, c.Hash())
+		}
 
-			keys := make([]multihash.Multihash, 0, len(m))
-			for c := range m {
-				delete(m, c)
+		// in case after removing all the invalid CIDs there are no valid ones left
+		if len(keys) == 0 {
+			continue
+		}
+		s.waitUntilProvideSystemReady()
+
+		log.Debugf("starting provide of %d keys", len(keys))
+		start := time.Now()
+		err := doProvideMany(s.ctx, s.rsys, keys)
+		if err != nil {
+			log.Debugf("providing failed %v", err)
+			continue
+		}
+		dur := time.Since(start)
+		recentAvgProvideDuration := dur / time.Duration(len(keys))
+		log.Debugf("finished providing of %d keys. It took %v with an average of %v per provide", len(keys), dur, recentAvgProvideDuration)
+	}
+}
 
-				// hash security
-				if err := verifcid.ValidateCid(s.allowlist, c); err != nil {
-					log.Errorf("insecure hash in reprovider, %s (%s)", c, err)
-					continue
-				}
+func (s *reprovider) instantProvideWorker() {
+	defer s.closewg.Done()
+	provCh := s.q.Dequeue()
 
-				keys = append(keys, c.Hash())
-			}
+	provideFunc := func(ctx context.Context, c cid.Cid) {
+		if err := s.rsys.Provide(ctx, c, true); err != nil {
+			log.Errorf("failed to provide %s: %s", c, err)
+		}
+	}
 
-			// in case after removing all the invalid CIDs there are no valid ones left
-			if len(keys) == 0 {
-				continue
+	var provideOperation func(context.Context, cid.Cid)
+	if s.provideWorkerCount == 0 {
+		// Unlimited workers
+		provideOperation = func(ctx context.Context, c cid.Cid) {
+			go provideFunc(ctx, c)
+		}
+	} else {
+		// Assign cid to workers pool
+		provideQueue := make(chan cid.Cid)
+		provideDelayTimer := time.NewTimer(provideDelayWarnDuration)
+		provideDelayTimer.Stop()
+		lateOnProvides := false
+
+		provideOperation = func(ctx context.Context, c cid.Cid) {
+			provideDelayTimer.Reset(provideDelayWarnDuration)
+			defer provideDelayTimer.Stop()
+			select {
+			case provideQueue <- c:
+				if lateOnProvides {
+					log.Warn("New provides are being processed again")
+				}
+				lateOnProvides = false
+			case <-provideDelayTimer.C:
+				if !lateOnProvides {
+					log.Warn("New provides are piling up in the queue, consider increasing the number of provide workers or enabling Batch Providing.")
```

Comment:
Instead of just logging a warning, this may be good place to increase the number of workers in the pool. That implies the pool has a starting size, and a maximum size. I am not sure it is worth the additional complexity to have a dynamically sized pool.

Reply:
The goal is to limit resource usage by default, and avoid growing to an unbounded number of workers. If user is fine with an unbounded number of workers and wants performance they can set `provideWorkerCount` to `0`, which is equivalent to increasing the number of workers as needed.

But maybe there is a better way to signal the user that they hit the limit? e.g user defined callback

Reply:
I was thinking of only growing worker pool up to its max size, not having an unbounded size. If worker pool if a fixed size, then his does not matter.